mlflow/models/evaluation/base.py

```diff
@@ -346,7 +377,26 @@ def hash(self):
                 column_names = ",".join(self.data.columns)
                 meta_str = f"columns={column_names}\nlabels={self.labels}"
                 md5_gen.update(meta_str.encode("UTF-8"))
-                EvaluationDataset._gen_md5_for_arraylike_obj(md5_gen, self.data)
+
+                data_for_hash = self.data
+
+                if self._spark_df_type and isinstance(self._original_data, self._spark_df_type):
+                    # For spark dataframe, the Vector type column we need expand it
+                    # into multiple columns, otherwise pandas hash function cannot compute hash
+                    # over it.
+                    transform_func = {}
+                    for field in self._original_data.schema:
+                        if isinstance(field.dataType, self._spark_vector_type):
+                            transform_func[field.name] = lambda x: pd.Series(x.toArray())
+                        else:
+                            transform_func[field.name] = lambda x: x
+
+                    data_for_hash = self.data.transform(transform_func)
+
+                # TODO:
+                #  For array/list type column values in pandas DataFrame, pandas hash function
+                #  also cannot support it, expand them if we need support them.
```

Comment:
Can we document this limitation in the `mlflow.evaluate()` API docs? What's the behavior right now if someone provides a pandas dataframe with array / list columns? Does it fail?

Reply:
> if someone provides a pandas dataframe with array / list columns

* In dataset hash compuation, error will raised. (The pandas hash function does not support it)
* I think no sklearn model support such input. Anyone can provide a counterexample ？

Reply:
@WeichenXu123 The pyfunc representation of TensorFlow models can accept pandas dataframes with list columns. While this is a rare case, I think it would be nice if we didn't fail. Is it easy to fix the hash? If not, can we leave the hash empty, fall back to using the Python object ID as the dataset name, and print a warning? I know the object ID changes each time, but I think this is better than failing.

Reply:
Having an object ID changing each time defeats the purpose of tracking the dataset. I'm more concerned with supporting so many different kinds of input. Is it really necessary to do so? Since it's fairly easy to convert among ndarray/Panda/SparkDF, I feel we are making the implementation unnecessarily complicated by trying to support these within the API instead of leaving it to the user.